polar-core/src/polar.rs

```diff
@@ -36,6 +37,19 @@ impl Query {
     pub fn next_message(&self) -> Option<Message> {
         self.vm.messages.next()
     }
+
+    pub fn source_info(&self) -> String {
+        let mut source_info = self.vm.term_source_string(&self.term);
+        if let Some(source) = self.vm.source(&self.term) {
```

Comment:
It still seems suboptimal that we're calling `self.vm.term_source_string(&self.term)`, which calls `self.vm.source(&self.term)`, and then we're calling `self.vm.source(&self.term)` _again_